megatron/model/transformer.py

```diff
@@ -533,6 +589,25 @@ def __init__(self, init_method,
         else:
             self.core_attention = MultiQueryCoreAttention(self.layer_number, self.attn_mask_type)
         self.checkpoint_core_attention = args.recompute_granularity == 'selective'
+        
+        if self.use_flash_attn:
+            if flash_attn_unpadded_func is None:
+                raise ImportError('FlashAttention is not installed, please install with '
+                                  'pip install flash-attn')
+            assert attention_type == AttnType.self_attn, ('FlashAttention code path only supports '
+                                                          'self-attention for now')
+            assert self.attn_mask_type == AttnMaskType.causal, ('FlashAttention code path only '
+                                                                'supports causal mask for now')
+            assert args.position_embedding_type != PositionEmbeddingType.alibi, \
+                ('FlashAttention does not support alibi positional embeddings yet')
+            if rearrange is None:
+                raise ImportError('einops is not installed, please install with pip install einops')
+            
+            if self.checkpoint_core_attention:
```

Comment:
I think this should raise an error. Also, selective recomputation has a huge performance impact so we really want it. How hard would it be to implement for FlashAttention?

Reply:
I'm not sure how hard it would be. Do you think applying the same `_checkpointed_attention_forward` code would work in this case?

Reply:
Also, do you think that even with flash-attention, selective recomputation would give a significant memory reduction compared to no-recomputation?

Reply:
Hmm maybe you're right, the goal of selective recomputation is to avoid storing the part that's quadratic in sequence length, but FlashAttention is linear

Reply:
(So we can change back to warning)

Reply:
Oh, nice! Then I'll change back to warning, yes